prover/man/man4/formcheck.4
.TH FORMCHECK 4 19/03/2001
.SH NAME
formcheck \- Check a formula with respect to a muCRL data specification

.SH SYNOPSIS
.B formcheck
[-formula 
.I 
form-file\c
] [-spec 
.I 
spec.tbf\c
] [OPTIONS]

.SH OPTIONS 
.IP \-print
prints the resulting BDDs
.IP \-counter
prints counter examples computed from resulting BDDs
.IP \-witness
prints witnesses computed from resulting BDDs
.IP \-help
Prints this help message

.in 0.75i
At least one of -formula or -spec must be present. If the other one
is missing, 
.I stdout
is taken as a default.

.SH DESCRIPTION
.B formcheck
is a tool that checks the formulae in 
.I form-file
using the algebraic data specification in
.I spec.tbf. 
It prints a list of "Yes", "No", or "Don't know" answers.  With the
-print, -counter and -witness options, one gets some diagnostic
information in case of a "Don't know" answer.

.SH FILES
.IP form-file
A file containing a list of variable declarations and formulae.
This file is of the form [[v(var,sort)..],formula,[v(var,sort)..],formula..]
.IP spec.tbf
A LPO in tbf-format, generated by mcrl -tbfile -binary spec

.SH SEE ALSO

prover(4)

.SH AUTHOR
Jaco van de Pol <vdpol@cwi.nl>
.LP
Please send all complaints, comments and bug fixes to me. 


svc/ChangeLog
   SVC -- the SVC (Systems Validation Centre) file format library

   Copyright (C) 2000  Stichting Mathematisch Centrum, Amsterdam,
                       The  Netherlands

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

$Id: ChangeLog,v 1.1.1.1 2004/09/07 15:06:33 uid523 Exp $



2000/6/1
   * Version 1.0 
   * Initial release

2000/6/15 (izak@cwi.nl)
   * Version 1.1 
   * Bug fix: frequency of the escape sequence is not updated

2000/9/18 (izak@cwi.nl)
   * Version 1.2
   * Update conform update ATerm library 1.4.4
   * Added configure.in to release

2000/9/25
   * Release 1.0.3
   * Fixed bug in troff manuals causing manuals to start with a blank page
   * Fixed memory allocation bug in compressed streams
   * Cleaned up header files by removing standard includes

2000/9/27
   * Release 1.0.4
   * Fixed bug: buffer for states and labels was allocated too small

2000/10/18
   * Release 1.0.5
   * Fixed bug: the SVCaterm2... functions returned 0 instead of -1 to flag
     an unknown ATerm

2001/1/4
   * Release 1.0.6
   * Format version number incremented to 1.2 beta
   * Lempel Ziff token enlarged to three bytes
   * Gnu General Public License added


prover/src/formcheck.c
/* $Id: formcheck.c,v 1.3 2004/11/23 12:36:17 uid523 Exp $ */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include "aterm2.h"
#include "mcrl.h"
#include "prover.h"
#include <string.h>

#define MY_NAME "Formcheck"
// VERSION assumed to be defined with -DVERSION=\"...\"

static char COUNTER=0;
static char WITNESS=0;
static char PRINT=0;
static char DOT=0;
static char* form_file=0;
static char* lpo_file=0;

void version() {
  fprintf(stderr,"%s (version %s)\n",MY_NAME, VERSION);
  exit(0);
}

static void WarningHandler(const char *format, va_list args)
     {
     fprintf(stderr,"%s: ", MY_NAME);
     ATvfprintf(stderr, format, args);
     fprintf(stderr,"\n");
     }
     
static void ErrorHandler(const char *format, va_list args)
     {
     fprintf(stderr,"%s: ", MY_NAME);
     ATvfprintf(stderr, format, args);
     fprintf(stderr,"\n");
     exit(-1);
     }

void help() {
  ATfprintf(stderr,"%s\n",
	    "\nformcheck [-formula file] [-spec file.tbf] [options]");
  ATfprintf(stderr,"%s\n",
	    "-- checks whether formulas in 'file' hold given the datatypes in 'file.tbf'");
  ATfprintf(stderr,"%s\n",
	    "");
  ATfprintf(stderr,"%s\n",
	    "-formula file  : "
	    "contains variable declarations and formulae to be proved:");
  ATfprintf(stderr,"%s\n",
	    "                 [[v(x,sort),...],formula,...]");
  ATfprintf(stderr,"%s\n",
	    "-spec file.tbf  : "
	    "contains the specification in which the formulae are proved");
  ATfprintf(stderr,"%s\n",
	    "At least one of "
	    "-formula or -spec must be present;");
  ATfprintf(stderr,"%s\n",
	    "if -formula or -spec are missing, stdin is taken.");
  ATfprintf(stderr,"%s\n",
	    "");
  ATfprintf(stderr,"%s\n",
	    "OPTIONS:");
  ATfprintf(stderr,
	    "%s\n","-print         : "
	    "prints resulting BDDs as ASCII to stderr");
  ATfprintf(stderr,
	    "%s\n","-print-dot     : "
	    "prints resulting BDDs as dot graphs to stdout");
  ATfprintf(stderr,
	    "%s\n","-counter       : "
	    "prints counter-examples");
  ATfprintf(stderr,
	    "%s\n","-witness       : "
	    "prints witness for the formula");
  ATfprintf(stderr,"%s\n",
	    "-version       : "
	    "prints the version number");
  ATfprintf(stderr,"%s\n",
	    "-help          : "
	    "prints this message");
  exit(0);
}

void parse_argument_line(int argc,char *argv[]) {
  int i;
  
  for (i=1;i<argc;i++) {
    if (!strcmp(argv[i],"-counter"))
      COUNTER=1; 
    if (!strcmp(argv[i],"-witness"))
      WITNESS=1; 
    if (!strcmp(argv[i],"-print"))
      PRINT=1; 
    if (!strcmp(argv[i],"-print-dot"))
      DOT=1; 
    if (!strcmp(argv[i],"-formula"))
      form_file=argv[++i]; 
    if (!strcmp(argv[i],"-spec"))
      lpo_file=argv[++i];
    if (!strcmp(argv[i],"-help"))
      help(); 
    if (!strcmp(argv[i],"-version"))
      version();
  }
  if (!(form_file || lpo_file)) 
    ATerror("Specify at least one of -formula or -spec\n");
}

int main(int argc, char *argv[]) {
  ATerm bottomOfStack=NULL;
  ATerm spec=NULL, formula=NULL;
  static ATermList vars=NULL,list=NULL;
  int count = 0;
  FILE *lpo;
  FILE *form;

  ATinit(argc,argv, &bottomOfStack);
  ATsetWarningHandler(WarningHandler);
  ATsetErrorHandler(ErrorHandler);
  parse_argument_line(argc,argv);

  ATprotect(&spec);
  ATprotect(&formula);
  ATprotect((ATerm*)&vars);
  ATprotect((ATerm*)&list);

  if (!lpo_file) 
    lpo=stdin;
  else {
    lpo = fopen(lpo_file,"r");
    if (!lpo) {
      perror(lpo_file);
      exit(1);
    }
  }
  spec = ATreadFromFile(lpo);
  if (spec==NULL) exit(1);

  if (!ATmatch(spec,"spec2gen(<term>,<term>)",
	       &spec,NULL))
    ATerror("Not a muCRL (2nd generation) specification on standard input.\n");

  MCRLinitializeAdt(spec);
  ProverSetArguments(&argc,&argv);
  ProverInitialize();
  //  Init_prover(spec,argc,argv);

  if (!form_file)
    form=stdin;
  else {
    form = fopen(form_file,"r");
    if (!form) {
      perror(form_file);
      exit(1);
    }
  }
  list=(ATermList)ATreadFromFile(form);
  if (list==NULL) exit(1);
  if (ATgetType(list)!=AT_LIST)
    ATerror("The -formula file should contain a list of declarations/formulae\n");

  while (!ATisEmpty(list)) {
    ATfprintf(stderr,"%d: ",++count);
    vars = (ATermList)ATgetFirst(list);
    if (ATgetType(vars)!=AT_LIST)
      ATerror("Variable declaration list [v(x,sort),...] expected\n %t\n",vars);
    vars = parse_decls(vars);
    list = ATgetNext(list);
    Declare_vars(vars);
    formula = parse(ATgetFirst(list));
    if (Sort_of(formula)!=prBOOL)
      ATerror("Formula should be of type Bool\n");
    list = ATgetNext(list);
    
    formula = Prove(formula);
    if (formula == prTRUE)
      fprintf(stderr,"True\n");
    else if (formula == prFALSE)
      fprintf(stderr,"False\n");
    else {
      fprintf(stderr,"Don't know\n");
      if (PRINT) print_BDD(formula,stderr);
      if (DOT) dot_BDD(formula,stdout);
      if (COUNTER) print_example(formula,prFALSE,stderr);
      if (WITNESS) print_example(formula,prTRUE,stderr);
    }
  }
  exit(0);
}
